# **Task 1: Data Quality Assessment**

**Assessment of data quality and completeness in preparation for analysis**

---

The client provided KPMG with 3 datasets:

- CustomerDemographic
- CustomerAddress
- Transactions data in past 3 months



In [1]:
#import required libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import datetime as dt

plt.style.use('ggplot')

In [2]:
xls = pd.ExcelFile(r'C:\Users\user\OneDrive\Máy tính\Personal Project\KPMG\Task 1 - Data Quality Assessment\KPMG_VI_New_raw_data_update_final (1).xlsx')
#transactions = pd.read_excel(xls, sheet_name='Transactions', skiprows=1)
#newCustomerList = pd.read_excel(xls, sheet_name='NewCustomerList', skiprows=1)
#customerDemographic = pd.read_excel(xls, sheet_name='CustomerDemographic', skiprows=1)
#customerAddress = pd.read_excel(xls, sheet_name='CustomerAddress', skiprows=1)

sheet_names = xls.sheet_names[1:]
sheet_to_df_map = {}
for sheet_name in sheet_names:
    sheet_to_df_map[sheet_name] = xls.parse(sheet_name, skiprows=1)

dfs = []
for sheet_name, df in sheet_to_df_map.items():
    globals()[sheet_name] = df
    dfs.append(df)

In [3]:
sheet_to_df_map.keys()

dict_keys(['Transactions', 'NewCustomerList', 'CustomerDemographic', 'CustomerAddress'])

## **Exploring Transactions Dataset**

**- Understanding the big picture**

**Column Description**

There are 20000 rows and 13 columns in `Transactions` dataset

- `transaction_id`: dtype: int - context: ID of transaction
- `product_id`: dtype: int - context: ID of product
- `customer_id`: dtype: int - context: ID of customer
- `transaction_date`: dtype: datetime - context: Date of transaction
- `online_order`: dtype: category (binary) - context: Whether the customer orders product online or not
- `order_status`: dtype: category - context: Is the order approved or cancelled?
- `brand`: dtype: category - context: Brand of product
- `product_line`: dtype: category - context: What the product line does the product belong to?
- `product_class`: dtype: category - context: Class of product
- `product_size`: dtype: category - context: Size of product
- `list_price`: dtype: float - context: Listed price of the product
- `standard_cost`: dtype: float - context: Cost price
- `product_first_sold_date`: dtype: datetime - context: The product was sold for the first time on

In [4]:
Transactions.shape

(20000, 13)

In [5]:
Transactions.head(5)

transaction_id  product_id  customer_id transaction_date  online_order   
0               1           2         2950       2017-02-25           0.0  \
1               2           3         3120       2017-05-21           1.0   
2               3          37          402       2017-10-16           0.0   
3               4          88         3135       2017-08-31           0.0   
4               5          78          787       2017-10-01           1.0   

  order_status           brand product_line product_class product_size   
0     Approved           Solex     Standard        medium       medium  \
1     Approved   Trek Bicycles     Standard        medium        large   
2     Approved      OHM Cycles     Standard           low       medium   
3     Approved  Norco Bicycles     Standard        medium       medium   
4     Approved  Giant Bicycles     Standard        medium        large   

   list_price  standard_cost  product_first_sold_date  
0       71.49          53.62                  41245.0  
1     2091.47         388.92                  41701.0  
2     1793.43         248.82                  36361.0  
3     1198.46         381.10                  36145.0  
4     1765.30         709.48                  42226.0

In [6]:
Transactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   transaction_id           20000 non-null  int64         
 1   product_id               20000 non-null  int64         
 2   customer_id              20000 non-null  int64         
 3   transaction_date         20000 non-null  datetime64[ns]
 4   online_order             19640 non-null  float64       
 5   order_status             20000 non-null  object        
 6   brand                    19803 non-null  object        
 7   product_line             19803 non-null  object        
 8   product_class            19803 non-null  object        
 9   product_size             19803 non-null  object        
 10  list_price               20000 non-null  float64       
 11  standard_cost            19803 non-null  float64       
 12  product_first_sold_date  19803 n

In [7]:
Transactions.describe(include='all')

transaction_id   product_id   customer_id            transaction_date   
count     20000.000000  20000.00000  20000.000000                       20000  \
unique             NaN          NaN           NaN                         NaN   
top                NaN          NaN           NaN                         NaN   
freq               NaN          NaN           NaN                         NaN   
mean      10000.500000     45.36465   1738.246050  2017-07-01 14:08:05.280000   
min           1.000000      0.00000      1.000000         2017-01-01 00:00:00   
25%        5000.750000     18.00000    857.750000         2017-04-01 00:00:00   
50%       10000.500000     44.00000   1736.000000         2017-07-03 00:00:00   
75%       15000.250000     72.00000   2613.000000         2017-10-02 00:00:00   
max       20000.000000    100.00000   5034.000000         2017-12-30 00:00:00   
std        5773.647028     30.75359   1011.951046                         NaN   

        online_order order_status  brand product_line product_class   
count   19640.000000        20000  19803        19803         19803  \
unique           NaN            2      6            4             3   
top              NaN     Approved  Solex     Standard        medium   
freq             NaN        19821   4253        14176         13826   
mean        0.500458          NaN    NaN          NaN           NaN   
min         0.000000          NaN    NaN          NaN           NaN   
25%         0.000000          NaN    NaN          NaN           NaN   
50%         1.000000          NaN    NaN          NaN           NaN   
75%         1.000000          NaN    NaN          NaN           NaN   
max         1.000000          NaN    NaN          NaN           NaN   
std         0.500013          NaN    NaN          NaN           NaN   

       product_size    list_price  standard_cost  product_first_sold_date  
count         19803  20000.000000   19803.000000             19803.000000  
unique            3           NaN            NaN                      NaN  
top          medium           NaN            NaN                      NaN  
freq          12990           NaN            NaN                      NaN  
mean            NaN   1107.829449     556.046951             38199.776549  
min             NaN     12.010000       7.210000             33259.000000  
25%             NaN    575.270000     215.140000             35667.000000  
50%             NaN   1163.890000     507.580000             38216.000000  
75%             NaN   1635.300000     795.100000             40672.000000  
max             NaN   2091.470000    1759.850000             42710.000000  
std             NaN    582.825242     405.955660              2875.201110

**- Preparation (Clean Data)**

In [8]:
# Checking Null value
Transactions.isna().sum()

transaction_id               0
product_id                   0
customer_id                  0
transaction_date             0
online_order               360
order_status                 0
brand                      197
product_line               197
product_class              197
product_size               197
list_price                   0
standard_cost              197
product_first_sold_date    197
dtype: int64

=> There are missing values in 7 columns. They can be dropped or treated according to the nature of analysis

In [9]:
# Checking duplicated values
Transactions.duplicated().sum()

0

=> There are no duplicated values

In [10]:
# product_first_sold_date column
# Convert excel date to real date
Transactions['product_first_sold_date'] = pd.TimedeltaIndex(Transactions['product_first_sold_date'], unit='D') + pd.Timestamp(1899, 12, 30)

In [11]:
Transactions.columns

Index(['transaction_id', 'product_id', 'customer_id', 'transaction_date',
       'online_order', 'order_status', 'brand', 'product_line',
       'product_class', 'product_size', 'list_price', 'standard_cost',
       'product_first_sold_date'],
      dtype='object')

In [12]:
Transactions['online_order'].value_counts(dropna=False)

online_order
1.0    9829
0.0    9811
NaN     360
Name: count, dtype: int64

In [13]:
Transactions['online_order'] = Transactions['online_order'].fillna(method='bfill')

In [14]:
Transactions.loc[Transactions['brand'].isna(),:]

transaction_id  product_id  customer_id transaction_date  online_order   
136               137           0          431       2017-09-23           0.0  \
159               160           0         3300       2017-08-27           0.0   
366               367           0         1614       2017-03-10           0.0   
406               407           0         2559       2017-06-14           1.0   
676               677           0         2609       2017-07-02           0.0   
...               ...         ...          ...              ...           ...   
19340           19341           0          443       2017-12-26           1.0   
19383           19384           0         2407       2017-06-11           0.0   
19793           19794           0         2860       2017-01-13           0.0   
19859           19860           0         2468       2017-06-24           1.0   
19871           19872           0           61       2017-03-17           1.0   

      order_status brand product_line product_class product_size  list_price   
136       Approved   NaN          NaN           NaN          NaN     1942.61  \
159       Approved   NaN          NaN           NaN          NaN     1656.86   
366       Approved   NaN          NaN           NaN          NaN      850.89   
406       Approved   NaN          NaN           NaN          NaN      710.59   
676       Approved   NaN          NaN           NaN          NaN     1972.01   
...            ...   ...          ...           ...          ...         ...   
19340     Approved   NaN          NaN           NaN          NaN      744.54   
19383     Approved   NaN          NaN           NaN          NaN     1098.18   
19793     Approved   NaN          NaN           NaN          NaN      868.56   
19859     Approved   NaN          NaN           NaN          NaN     1497.43   
19871     Approved   NaN          NaN           NaN          NaN      867.92   

       standard_cost product_first_sold_date  
136              NaN                     NaT  
159              NaN                     NaT  
366              NaN                     NaT  
406              NaN                     NaT  
676              NaN                     NaT  
...              ...                     ...  
19340            NaN                     NaT  
19383            NaN                     NaT  
19793            NaN                     NaT  
19859            NaN                     NaT  
19871            NaN                     NaT  

[197 rows x 13 columns]

In [15]:
Transactions.dropna(inplace=True)

In [16]:
Transactions.reset_index(drop=True)

transaction_id  product_id  customer_id transaction_date  online_order   
0                   1           2         2950       2017-02-25           0.0  \
1                   2           3         3120       2017-05-21           1.0   
2                   3          37          402       2017-10-16           0.0   
3                   4          88         3135       2017-08-31           0.0   
4                   5          78          787       2017-10-01           1.0   
...               ...         ...          ...              ...           ...   
19798           19996          51         1018       2017-06-24           1.0   
19799           19997          41          127       2017-11-09           1.0   
19800           19998          87         2284       2017-04-14           1.0   
19801           19999           6         2764       2017-07-03           0.0   
19802           20000          11         1144       2017-09-22           1.0   

      order_status           brand product_line product_class product_size   
0         Approved           Solex     Standard        medium       medium  \
1         Approved   Trek Bicycles     Standard        medium        large   
2         Approved      OHM Cycles     Standard           low       medium   
3         Approved  Norco Bicycles     Standard        medium       medium   
4         Approved  Giant Bicycles     Standard        medium        large   
...            ...             ...          ...           ...          ...   
19798     Approved      OHM Cycles     Standard          high       medium   
19799     Approved           Solex         Road        medium       medium   
19800     Approved      OHM Cycles     Standard        medium       medium   
19801     Approved      OHM Cycles     Standard          high       medium   
19802     Approved   Trek Bicycles     Standard        medium        small   

       list_price  standard_cost product_first_sold_date  
0           71.49          53.62              2012-12-02  
1         2091.47         388.92              2014-03-03  
2         1793.43         248.82              1999-07-20  
3         1198.46         381.10              1998-12-16  
4         1765.30         709.48              2015-08-10  
...           ...            ...                     ...  
19798     2005.66        1203.40              2003-07-21  
19799      416.98         312.74              1997-05-10  
19800     1636.90          44.71              2010-08-20  
19801      227.88         136.73              2004-08-17  
19802     1775.81        1580.47              1999-06-23  

[19803 rows x 13 columns]

In [17]:
cols = ['online_order','order_status','brand','product_line','product_class','product_size']
for col in cols:
    print(f"\n --- {col} ---")
    print(Transactions[col].value_counts(dropna=False))


 --- online_order ---
online_order
1.0    9946
0.0    9857
Name: count, dtype: int64

 --- order_status ---
order_status
Approved     19625
Cancelled      178
Name: count, dtype: int64

 --- brand ---
brand
Solex             4253
Giant Bicycles    3312
WeareA2B          3295
OHM Cycles        3043
Trek Bicycles     2990
Norco Bicycles    2910
Name: count, dtype: int64

 --- product_line ---
product_line
Standard    14176
Road         3970
Touring      1234
Mountain      423
Name: count, dtype: int64

 --- product_class ---
product_class
medium    13826
high       3013
low        2964
Name: count, dtype: int64

 --- product_size ---
product_size
medium    12990
large      3976
small      2837
Name: count, dtype: int64


In [18]:
Transactions.to_csv('Transactions.csv', index=False)

## **Exploring NewCustomerList Dataset**

**- Understanding the big picture**

**Column Description**

There are 1000 rows and 18 columns in `NewCustomerList` dataset

- `first_name`: dtype: string - context: First name of customer
- `last_name`: dtype: string - context: Last name of customer
- `gender`: dtype: category - context: Gender of customer
- `past_3_years_bike_related_purchases`: dtype: float - context: Bicycle-related purchases in the past 3 years
- `DOB`: dtype: datetime - context: Date of birth of customer
- `job_title`: dtype: category - context: Job title of customer
- `job_industry_category`: dtype: category - context: Job industry category of customer
- `wealth_segment`: dtype: category - context: Wealth segment of customer
- `deceased_indicator`: dtype: category - context: Is the customer still alive?
- `owns_car`: dtype category (binary) - context: Does the customer own a car?
- `tenure`: dtype: int - context: Number of years since the customer became a customer of the company
- `address`: dtype: string - context: Address of customer
- `postcode`: dtype: category - context: Postcode of city that the customer lives
- `state`: dtype: category - context: State where the customer lives
- `country`: dtype: category - context: Country where the customer lives
- `property_valuation`: dtype: int - context: Property valuation of customer
- `Rank`: dtype: int - context: Customer ranking
- `Value`: dtype: float - context

In [19]:
NewCustomerList.head()

first_name  last_name  gender  past_3_years_bike_related_purchases   
0    Chickie    Brister    Male                                   86  \
1      Morly     Genery    Male                                   69   
2    Ardelis  Forrester  Female                                   10   
3     Lucine      Stutt  Female                                   64   
4    Melinda     Hadlee  Female                                   34   

                   DOB                   job_title job_industry_category   
0           1957-07-12             General Manager         Manufacturing  \
1           1970-03-22         Structural Engineer              Property   
2  1974-08-28 00:00:00      Senior Cost Accountant    Financial Services   
3           1979-01-28  Account Representative III         Manufacturing   
4           1965-09-21           Financial Analyst    Financial Services   

      wealth_segment deceased_indicator owns_car  ...  state    country   
0      Mass Customer                  N      Yes  ...    QLD  Australia  \
1      Mass Customer                  N       No  ...    NSW  Australia   
2  Affluent Customer                  N       No  ...    VIC  Australia   
3  Affluent Customer                  N      Yes  ...    QLD  Australia   
4  Affluent Customer                  N       No  ...    NSW  Australia   

   property_valuation Unnamed: 16 Unnamed: 17  Unnamed: 18  Unnamed: 19   
0                   6        1.10       1.375      1.71875     1.460938  \
1                  11        0.93       0.930      1.16250     0.988125   
2                   5        0.47       0.470      0.47000     0.470000   
3                   1        1.06       1.325      1.32500     1.325000   
4                   9        0.57       0.570      0.71250     0.712500   

   Unnamed: 20  Rank     Value  
0            1     1  1.718750  
1            1     1  1.718750  
2            1     1  1.718750  
3            4     4  1.703125  
4            4     4  1.703125  

[5 rows x 23 columns]

In [20]:
NewCustomerList.drop(columns=['Unnamed: 16','Unnamed: 17','Unnamed: 18','Unnamed: 19', 'Unnamed: 20'],
                     axis=1,inplace=True)

In [21]:
NewCustomerList.shape

(1000, 18)

In [22]:
NewCustomerList.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 18 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   first_name                           1000 non-null   object 
 1   last_name                            971 non-null    object 
 2   gender                               1000 non-null   object 
 3   past_3_years_bike_related_purchases  1000 non-null   int64  
 4   DOB                                  983 non-null    object 
 5   job_title                            894 non-null    object 
 6   job_industry_category                835 non-null    object 
 7   wealth_segment                       1000 non-null   object 
 8   deceased_indicator                   1000 non-null   object 
 9   owns_car                             1000 non-null   object 
 10  tenure                               1000 non-null   int64  
 11  address                        

In [23]:
NewCustomerList.describe(include='all')

first_name last_name  gender  past_3_years_bike_related_purchases   
count        1000       971    1000                          1000.000000  \
unique        940       961       3                                  NaN   
top      Rozamond    Sissel  Female                                  NaN   
freq            3         2     513                                  NaN   
mean          NaN       NaN     NaN                            49.836000   
std           NaN       NaN     NaN                            27.796686   
min           NaN       NaN     NaN                             0.000000   
25%           NaN       NaN     NaN                            26.750000   
50%           NaN       NaN     NaN                            51.000000   
75%           NaN       NaN     NaN                            72.000000   
max           NaN       NaN     NaN                            99.000000   

               DOB            job_title job_industry_category wealth_segment   
count          983                  894                   835           1000  \
unique         961                  184                     9              3   
top     1965-07-03  Associate Professor    Financial Services  Mass Customer   
freq             2                   15                   203            508   
mean           NaN                  NaN                   NaN            NaN   
std            NaN                  NaN                   NaN            NaN   
min            NaN                  NaN                   NaN            NaN   
25%            NaN                  NaN                   NaN            NaN   
50%            NaN                  NaN                   NaN            NaN   
75%            NaN                  NaN                   NaN            NaN   
max            NaN                  NaN                   NaN            NaN   

       deceased_indicator owns_car       tenure           address   
count                1000     1000  1000.000000              1000  \
unique                  1        2          NaN              1000   
top                     N       No          NaN  45 Shopko Center   
freq                 1000      507          NaN                 1   
mean                  NaN      NaN    11.388000               NaN   
std                   NaN      NaN     5.037145               NaN   
min                   NaN      NaN     0.000000               NaN   
25%                   NaN      NaN     7.000000               NaN   
50%                   NaN      NaN    11.000000               NaN   
75%                   NaN      NaN    15.000000               NaN   
max                   NaN      NaN    22.000000               NaN   

           postcode state    country  property_valuation         Rank   
count   1000.000000  1000       1000         1000.000000  1000.000000  \
unique          NaN     3          1                 NaN          NaN   
top             NaN   NSW  Australia                 NaN          NaN   
freq            NaN   506       1000                 NaN          NaN   
mean    3019.227000   NaN        NaN            7.397000   498.819000   
std      848.895767   NaN        NaN            2.758804   288.810997   
min     2000.000000   NaN        NaN            1.000000     1.000000   
25%     2209.000000   NaN        NaN            6.000000   250.000000   
50%     2800.000000   NaN        NaN            8.000000   500.000000   
75%     3845.500000   NaN        NaN            9.000000   750.250000   
max     4879.000000   NaN        NaN           12.000000  1000.000000   

              Value  
count   1000.000000  
unique          NaN  
top             NaN  
freq            NaN  
mean       0.881714  
std        0.293525  
min        0.340000  
25%        0.649531  
50%        0.860000  
75%        1.075000  
max        1.718750

**- Preparation (Clean Data)**

In [24]:
NewCustomerList.isna().sum()

first_name                               0
last_name                               29
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     17
job_title                              106
job_industry_category                  165
wealth_segment                           0
deceased_indicator                       0
owns_car                                 0
tenure                                   0
address                                  0
postcode                                 0
state                                    0
country                                  0
property_valuation                       0
Rank                                     0
Value                                    0
dtype: int64

=> There are missing values in 4 columns

In [25]:
NewCustomerList.duplicated().sum()

0

=> There are no duplicated values

In [26]:
NewCustomerList.columns

Index(['first_name', 'last_name', 'gender',
       'past_3_years_bike_related_purchases', 'DOB', 'job_title',
       'job_industry_category', 'wealth_segment', 'deceased_indicator',
       'owns_car', 'tenure', 'address', 'postcode', 'state', 'country',
       'property_valuation', 'Rank', 'Value'],
      dtype='object')

In [27]:
NewCustomerList['gender'].value_counts(dropna=False)

gender
Female    513
Male      470
U          17
Name: count, dtype: int64

In [28]:
NewCustomerList.query("gender=='U'")

first_name    last_name gender  past_3_years_bike_related_purchases  DOB   
59       Normy     Goodinge      U                                    5  NaN  \
226      Hatti     Carletti      U                                   35  NaN   
324   Rozamond       Turtle      U                                   69  NaN   
358      Tamas      Swatman      U                                   65  NaN   
360      Tracy   Andrejevic      U                                   71  NaN   
374     Agneta      McAmish      U                                   66  NaN   
434      Gregg      Aimeric      U                                   52  NaN   
439      Johna       Bunker      U                                   93  NaN   
574    Harlene         Nono      U                                   69  NaN   
598   Gerianne       Kaysor      U                                   15  NaN   
664     Chicky      Sinclar      U                                   43  NaN   
751    Adriana  Saundercock      U                                   20  NaN   
775     Dmitri        Viant      U                                   62  NaN   
835      Porty       Hansed      U                                   88  NaN   
883      Shara     Bramhill      U                                   24  NaN   
904       Roth         Crum      U                                    0  NaN   
984    Pauline     Dallosso      U                                   82  NaN   

                        job_title job_industry_category     wealth_segment   
59            Associate Professor                    IT      Mass Customer  \
226               Legal Assistant                    IT  Affluent Customer   
324               Legal Assistant                    IT      Mass Customer   
358       Assistant Media Planner         Entertainment  Affluent Customer   
360                 Programmer II                    IT      Mass Customer   
374  Structural Analysis Engineer                    IT      Mass Customer   
434              Internal Auditor                    IT      Mass Customer   
439                Tax Accountant                    IT      Mass Customer   
574       Human Resources Manager                    IT      Mass Customer   
598               Project Manager                    IT  Affluent Customer   
664                      Operator                    IT     High Net Worth   
751                         Nurse                    IT     High Net Worth   
775                     Paralegal    Financial Services  Affluent Customer   
835               General Manager                    IT      Mass Customer   
883                           NaN                    IT  Affluent Customer   
904               Legal Assistant                    IT      Mass Customer   
984    Desktop Support Technician                    IT  Affluent Customer   

    deceased_indicator owns_car  tenure                  address  postcode   
59                   N       No       4      7232 Fulton Parkway      3810  \
226                  N      Yes      11            6 Iowa Center      2519   
324                  N      Yes       3  57025 New Castle Street      3850   
358                  N       No       5       78 Clarendon Drive      4551   
360                  N      Yes      11  5675 Burning Wood Trail      3030   
374                  N       No      15           5773 Acker Way      4207   
434                  N       No       7      72423 Surrey Street      3753   
439                  N      Yes      14         3686 Waubesa Way      3065   
574                  N       No      12  0307 Namekagon Crossing      2170   
598                  N       No       5           882 Toban Lane      2121   
664                  N      Yes       0        5 Red Cloud Place      3222   
751                  N      Yes      14         82 Gina Junction      3806   
775                  N       No       5     95960 Warner Parkway      3842   
835                  N       No      13     768 Southridge Drive  

=> There are 17 rows with unkonwn/unspecified gender

In [29]:
NewCustomerList['gender'] = NewCustomerList['gender'].replace('U','Unspecifeid')

In [30]:
NewCustomerList['last_name'] = NewCustomerList['last_name'].fillna(method='bfill')

In [31]:
NewCustomerList['DOB'].value_counts(dropna=False)

DOB
NaN           17
1965-07-03     2
1978-01-15     2
1979-07-28     2
1995-08-13     2
              ..
1978-05-27     1
1945-08-08     1
1943-08-27     1
1999-10-24     1
1955-10-02     1
Name: count, Length: 962, dtype: int64

In [32]:
NewCustomerList.loc[NewCustomerList['DOB'].isna(),:]

first_name    last_name       gender  past_3_years_bike_related_purchases   
59       Normy     Goodinge  Unspecifeid                                    5  \
226      Hatti     Carletti  Unspecifeid                                   35   
324   Rozamond       Turtle  Unspecifeid                                   69   
358      Tamas      Swatman  Unspecifeid                                   65   
360      Tracy   Andrejevic  Unspecifeid                                   71   
374     Agneta      McAmish  Unspecifeid                                   66   
434      Gregg      Aimeric  Unspecifeid                                   52   
439      Johna       Bunker  Unspecifeid                                   93   
574    Harlene         Nono  Unspecifeid                                   69   
598   Gerianne       Kaysor  Unspecifeid                                   15   
664     Chicky      Sinclar  Unspecifeid                                   43   
751    Adriana  Saundercock  Unspecifeid                                   20   
775     Dmitri        Viant  Unspecifeid                                   62   
835      Porty       Hansed  Unspecifeid                                   88   
883      Shara     Bramhill  Unspecifeid                                   24   
904       Roth         Crum  Unspecifeid                                    0   
984    Pauline     Dallosso  Unspecifeid                                   82   

     DOB                     job_title job_industry_category   
59   NaN           Associate Professor                    IT  \
226  NaN               Legal Assistant                    IT   
324  NaN               Legal Assistant                    IT   
358  NaN       Assistant Media Planner         Entertainment   
360  NaN                 Programmer II                    IT   
374  NaN  Structural Analysis Engineer                    IT   
434  NaN              Internal Auditor                    IT   
439  NaN                Tax Accountant                    IT   
574  NaN       Human Resources Manager                    IT   
598  NaN               Project Manager                    IT   
664  NaN                      Operator                    IT   
751  NaN                         Nurse                    IT   
775  NaN                     Paralegal    Financial Services   
835  NaN               General Manager                    IT   
883  NaN                           NaN                    IT   
904  NaN               Legal Assistant                    IT   
984  NaN    Desktop Support Technician                    IT   

        wealth_segment deceased_indicator owns_car  tenure   
59       Mass Customer                  N       No       4  \
226  Affluent Customer                  N      Yes      11   
324      Mass Customer                  N      Yes       3   
358  Affluent Customer                  N       No       5   
360      Mass Customer                  N      Yes      11   
374      Mass Customer                  N       No      15   
434      Mass Customer                  N       No       7   
439      Mass Customer                  N      Yes      14   
574      Mass Customer                  N       No      12   
598  Affluent Customer                  N       No       5   
664     High Net Worth                  N      Yes       0   
751     High Net Worth                  N      Yes      14   
775  Affluent Customer                  N       No       5   
835      Mass Customer                  N       No      13   
883  Affluent Customer                  N       No       2   
904      Mass Customer                  N       No       2   
984  Affluent Customer                  N      Yes       0   

                     address  postcode state    country  property_valuation   
59       7232 Fulton Parkway      3810   VIC  Australia                   5  \
226            6 Iowa Center      2519   NSW  Australia                   9   
324  57025 New Castle Street      3850

In [33]:
NewCustomerList['job_title'].value_counts(dropna=False)

job_title
NaN                      106
Associate Professor       15
Software Consultant       14
Environmental Tech        14
Chief Design Engineer     13
                        ... 
Staff Accountant II        1
Web Developer II           1
Statistician III           1
Programmer Analyst IV      1
Web Developer I            1
Name: count, Length: 185, dtype: int64

In [34]:
NewCustomerList['job_industry_category'].value_counts(dropna=False)

job_industry_category
Financial Services    203
Manufacturing         199
NaN                   165
Health                152
Retail                 78
Property               64
IT                     51
Entertainment          37
Argiculture            26
Telecommunications     25
Name: count, dtype: int64

In [35]:
NewCustomerList.loc[NewCustomerList['job_title'].isna(),:]

first_name  last_name  gender  past_3_years_bike_related_purchases   
15       Dukie      Swire    Male                                   88  \
25      Rourke   Gillbard    Male                                   11   
29       Rhona  De Freyne  Female                                   45   
30     Sharron  Claibourn  Female                                   62   
37    Mitchell   MacCague    Male                                   58   
..         ...        ...     ...                                  ...   
952      Candy      Cokly  Female                                   23   
953      Noami      Cokly  Female                                   74   
971     Frieda    Tavinor  Female                                   43   
972    Ellwood     Budden    Male                                   82   
989     Kellen   Pawelski  Female                                   83   

            DOB job_title job_industry_category     wealth_segment   
15   1954-03-31       NaN         Manufacturing  Affluent Customer  \
25   1945-08-03       NaN              Property      Mass Customer   
29   1960-11-22       NaN                Health     High Net Worth   
30   1980-01-26       NaN    Financial Services     High Net Worth   
37   1979-04-11       NaN         Manufacturing      Mass Customer   
..          ...       ...                   ...                ...   
952  1977-12-08       NaN    Financial Services      Mass Customer   
953  1962-09-17       NaN         Manufacturing      Mass Customer   
971  1999-03-04       NaN                   NaN  Affluent Customer   
972  1998-06-03       NaN                Health      Mass Customer   
989  1945-07-26       NaN         Manufacturing     High Net Worth   

    deceased_indicator owns_car  tenure                    address  postcode   
15                   N      Yes       5          64 Granby Parkway      2500  \
25                   N       No      17          75 Cordelia Trail      4817   
29                   N       No       8           11184 East Drive      3056   
30                   N      Yes      17         555 Hermina Avenue      2280   
37                   N       No      15           240 Acker Avenue      3190   
..                 ...      ...     ...                        ...       ...   
952                  N       No       6       59252 Maryland Drive      3500   
953                  N      Yes      15   2886 Buena Vista Terrace      2038   
971                  N       No      10             7 Mallory Lane      3064   
972                  N      Yes      11         79907 Randy Center      2192   
989                  N      Yes      11  125 Manufacturers Parkway      2193   

    state    country  property_valuation  Rank     Value  
15    NSW  Australia                   8    16  1.562500  
25    QLD  Australia                   4    26  1.468750  
29    VIC  Australia                  10    30  1.460938  
30    NSW  Australia                   8    30  1.460938  
37    VIC  Australia                   8    38  1.437500  
..    ...        ...                 ...   ...       ...  
952   VIC  Australia                   3   951  0.450500  
953   NSW  Australia                  11   954  0.450000  
971   VIC  Australia                   6   972  0.430000  
972   NSW  Australia                  10   972  0.430000  
989   NSW  Australia                   8   988  0.399500  

[106 rows x 18 columns]

In [36]:
NewCustomerList['wealth_segment'].value_counts(dropna=False)

wealth_segment
Mass Customer        508
High Net Worth       251
Affluent Customer    241
Name: count, dtype: int64

In [37]:
NewCustomerList['state'].value_counts(dropna=False)

state
NSW    506
VIC    266
QLD    228
Name: count, dtype: int64

In [38]:
NewCustomerList['owns_car'].value_counts(dropna=False)

owns_car
No     507
Yes    493
Name: count, dtype: int64

In [39]:
NewCustomerList.to_csv('NewCustomerList.csv')

## **Exploring CustomerDemographic Dataset**

**- Understanding the big picture**

**Column Description**

There are 4000 rows and 12 columns in `CustomerDemographic` dataset

- `first_name`: dtype: string - context: First name of customer
- `last_name`: dtype: string - context: Last name of customer
- `gender`: dtype: category - context: Gender of customer
- `past_3_years_bike_related_purchases`: dtype: float - context: Bicycle-related purchases in the past 3 years
- `DOB`: dtype: datetime - context: Date of birth of customer
- `job_title`: dtype: category - context: Job title of customer
- `job_industry_category`: dtype: category - context: Job industry category of customer
- `wealth_segment`: dtype: category - context: Wealth segment of customer
- `deceased_indicator`: dtype: category - context: Is the customer still alive?
- `owns_car`: dtype category (binary) - context: Does the customer own a car?
- `tenure`: dtype: int - context: Number of years since the customer became a customer of the company
- `address`: dtype: string - context: Address of customer

In [40]:
CustomerDemographic.head()

customer_id      first_name  last_name  gender   
0            1         Laraine  Medendorp       F  \
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot        NaN    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases                  DOB   
0                                   93  1953-10-12 00:00:00  \
1                                   81  1980-12-16 00:00:00   
2                                   61  1954-01-20 00:00:00   
3                                   33  1961-10-03 00:00:00   
4                                   56  1977-05-13 00:00:00   

                job_title job_industry_category     wealth_segment   
0     Executive Secretary                Health      Mass Customer  \
1  Administrative Officer    Financial Services      Mass Customer   
2      Recruiting Manager              Property      Mass Customer   
3                     NaN                    IT      Mass Customer   
4           Senior Editor                   NaN  Affluent Customer   

  deceased_indicator                                            default   
0                  N                                                 "'  \
1                  N                       <script>alert('hi')</script>   
2                  N                                2018-02-01 00:00:00   
3                  N  () { _; } >_[$($())] { touch /tmp/blns.shellsh...   
4                  N                                                NIL   

  owns_car  tenure  
0      Yes    11.0  
1      Yes    16.0  
2      Yes    15.0  
3       No     7.0  
4      Yes     8.0

In [41]:
# The values in default column are inconsistent, hence dropping the column 
CustomerDemographic.drop('default', axis=1, inplace=True)

In [42]:
CustomerDemographic.shape

(4000, 12)

In [43]:
CustomerDemographic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 12 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   customer_id                          4000 non-null   int64  
 1   first_name                           4000 non-null   object 
 2   last_name                            3875 non-null   object 
 3   gender                               4000 non-null   object 
 4   past_3_years_bike_related_purchases  4000 non-null   int64  
 5   DOB                                  3913 non-null   object 
 6   job_title                            3494 non-null   object 
 7   job_industry_category                3344 non-null   object 
 8   wealth_segment                       4000 non-null   object 
 9   deceased_indicator                   4000 non-null   object 
 10  owns_car                             4000 non-null   object 
 11  tenure                        

In [44]:
CustomerDemographic.describe(include='all')

customer_id first_name last_name  gender   
count   4000.000000       4000      3875    4000  \
unique          NaN       3139      3725       6   
top             NaN        Max  Pristnor  Female   
freq            NaN          5         3    2037   
mean    2000.500000        NaN       NaN     NaN   
std     1154.844867        NaN       NaN     NaN   
min        1.000000        NaN       NaN     NaN   
25%     1000.750000        NaN       NaN     NaN   
50%     2000.500000        NaN       NaN     NaN   
75%     3000.250000        NaN       NaN     NaN   
max     4000.000000        NaN       NaN     NaN   

        past_3_years_bike_related_purchases                  DOB   
count                           4000.000000                 3913  \
unique                                  NaN                 3448   
top                                     NaN  1978-01-30 00:00:00   
freq                                    NaN                    7   
mean                              48.890000                  NaN   
std                               28.715005                  NaN   
min                                0.000000                  NaN   
25%                               24.000000                  NaN   
50%                               48.000000                  NaN   
75%                               73.000000                  NaN   
max                               99.000000                  NaN   

                                   job_title job_industry_category   
count                                   3494                  3344  \
unique                                   195                     9   
top     Business Systems Development Analyst         Manufacturing   
freq                                      45                   799   
mean                                     NaN                   NaN   
std                                      NaN                   NaN   
min                                      NaN                   NaN   
25%                                      NaN                   NaN   
50%                                      NaN                   NaN   
75%                                      NaN                   NaN   
max                                      NaN                   NaN   

       wealth_segment deceased_indicator owns_car       tenure  
count            4000               4000     4000  3913.000000  
unique              3                  2        2          NaN  
top     Mass Customer                  N      Yes          NaN  
freq             2000               3998     2024          NaN  
mean              NaN                NaN      NaN    10.657041  
std               NaN                NaN      NaN     5.660146  
min               NaN                NaN      NaN     1.000000  
25%               NaN                NaN      NaN     6.000000  
50%               NaN                NaN      NaN    11.000000  
75%               NaN                NaN      NaN    15.000000  
max               NaN                NaN      NaN    22.000000

**- Preparation (Clean data)**

In [45]:
CustomerDemographic.isna().sum()

customer_id                              0
first_name                               0
last_name                              125
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     87
job_title                              506
job_industry_category                  656
wealth_segment                           0
deceased_indicator                       0
owns_car                                 0
tenure                                  87
dtype: int64

=> There are missing values in 5 columns

In [46]:
CustomerDemographic.duplicated().sum()

0

=> There are no duplicated values

In [47]:
CustomerDemographic.columns

Index(['customer_id', 'first_name', 'last_name', 'gender',
       'past_3_years_bike_related_purchases', 'DOB', 'job_title',
       'job_industry_category', 'wealth_segment', 'deceased_indicator',
       'owns_car', 'tenure'],
      dtype='object')

In [48]:
CustomerDemographic['customer_id'].nunique()

4000

In [49]:
CustomerDemographic['last_name'] = CustomerDemographic['last_name'].fillna(method='bfill')

In [50]:
CustomerDemographic['gender'].value_counts(dropna=False)

gender
Female    2037
Male      1872
U           88
F            1
Femal        1
M            1
Name: count, dtype: int64

In [51]:
gender = ['Male','Female','Unspecified']

male = "Male|M"
female = "Female|F|Femal"
unspecifeid = "U"

conditions = [
    (CustomerDemographic['gender'].str.contains(male)),
    (CustomerDemographic['gender'].str.contains(female)),
    (CustomerDemographic['gender'].str.contains(unspecifeid))
    ]

CustomerDemographic['gender'] = np.select(conditions, gender, default='Other')

In [53]:
CustomerDemographic['job_title'].value_counts(dropna=False)

job_title
NaN                                     506
Business Systems Development Analyst     45
Tax Accountant                           44
Social Worker                            44
Internal Auditor                         42
                                       ... 
Database Administrator I                  4
Health Coach I                            3
Health Coach III                          3
Research Assistant III                    3
Developer I                               1
Name: count, Length: 196, dtype: int64

In [54]:
CustomerDemographic['job_industry_category'].value_counts(dropna=False)

job_industry_category
Manufacturing         799
Financial Services    774
NaN                   656
Health                602
Retail                358
Property              267
IT                    223
Entertainment         136
Argiculture           113
Telecommunications     72
Name: count, dtype: int64

In [86]:
CustomerDemographic['DOB'] = pd.to_datetime(CustomerDemographic['DOB'])

In [89]:
CustomerDemographic['DOB'].min() #This DOB seems impractical

Timestamp('1843-12-21 00:00:00')

In [90]:
CustomerDemographic.loc[CustomerDemographic['DOB']=='1843-12-21',:]

customer_id first_name last_name       gender   
33           34   Jephthah  Bachmann  Unspecified  \

    past_3_years_bike_related_purchases        DOB        job_title   
33                                   59 1843-12-21  Legal Assistant  \

   job_industry_category     wealth_segment deceased_indicator owns_car   
33                    IT  Affluent Customer                  N       No  \

    tenure  
33    20.0

In [55]:
#categories = CustomerDemographic['job_industry_category'].dropna().unique()
#CustomerDemographic['job_industry_category'] = CustomerDemographic['job_industry_category']\
#                                               .fillna(pd.Series(np.random.choice(categories, size=len(CustomerDemographic['job_industry_category']))))

In [57]:
CustomerDemographic['wealth_segment'].value_counts(dropna=False)

wealth_segment
Mass Customer        2000
High Net Worth       1021
Affluent Customer     979
Name: count, dtype: int64

In [58]:
CustomerDemographic['deceased_indicator'].value_counts(dropna=False)

deceased_indicator
N    3998
Y       2
Name: count, dtype: int64

In [59]:
CustomerDemographic['owns_car'].value_counts(dropna=False)

owns_car
Yes    2024
No     1976
Name: count, dtype: int64

## **Exploring CustomerAddress Dataset**

**- Understanding the big picture**

**Column Description**

There are 3999 rows and 6 columns in `CustomerAddress` dataset

- `customer_id`: dtype: int - context: ID of customer
- `address`: dtype: string - context: Address of customer
- `postcode`: dtype: category - context: Postcode of city that the customer lives
- `state`: dtype: category - context: State where the customer lives
- `country`: dtype: category - context: Country where the customer lives
- `property_valuation`: dtype: int - context: Property valuation of customer


In [60]:
CustomerAddress.head()

customer_id              address  postcode            state    country   
0            1   060 Morning Avenue      2016  New South Wales  Australia  \
1            2  6 Meadow Vale Court      2153  New South Wales  Australia   
2            4   0 Holy Cross Court      4211              QLD  Australia   
3            5  17979 Del Mar Point      2448  New South Wales  Australia   
4            6     9 Oakridge Court      3216              VIC  Australia   

   property_valuation  
0                  10  
1                  10  
2                   9  
3                   4  
4                   9

In [61]:
CustomerAddress.shape

(3999, 6)

In [62]:
CustomerAddress.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3999 entries, 0 to 3998
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   customer_id         3999 non-null   int64 
 1   address             3999 non-null   object
 2   postcode            3999 non-null   int64 
 3   state               3999 non-null   object
 4   country             3999 non-null   object
 5   property_valuation  3999 non-null   int64 
dtypes: int64(3), object(3)
memory usage: 187.6+ KB


In [63]:
CustomerAddress.describe(include='all')

customer_id                  address     postcode state    country   
count   3999.000000                     3999  3999.000000  3999       3999  \
unique          NaN                     3996          NaN     5          1   
top             NaN  3 Mariners Cove Terrace          NaN   NSW  Australia   
freq            NaN                        2          NaN  2054       3999   
mean    2003.987997                      NaN  2985.755939   NaN        NaN   
std     1154.576912                      NaN   844.878364   NaN        NaN   
min        1.000000                      NaN  2000.000000   NaN        NaN   
25%     1004.500000                      NaN  2200.000000   NaN        NaN   
50%     2004.000000                      NaN  2768.000000   NaN        NaN   
75%     3003.500000                      NaN  3750.000000   NaN        NaN   
max     4003.000000                      NaN  4883.000000   NaN        NaN   

        property_valuation  
count          3999.000000  
unique                 NaN  
top                    NaN  
freq                   NaN  
mean              7.514379  
std               2.824663  
min               1.000000  
25%               6.000000  
50%               8.000000  
75%              10.000000  
max              12.000000

**- Preparation (Clean data)**

In [64]:
CustomerAddress.isna().sum()

customer_id           0
address               0
postcode              0
state                 0
country               0
property_valuation    0
dtype: int64

=> There are no missing values

In [65]:
CustomerAddress.duplicated().sum()

0

=> There are no duplicated values

In [66]:
CustomerAddress['postcode'].value_counts(dropna=False)

postcode
2170    31
2155    30
2145    30
2153    29
3977    26
        ..
3808     1
3114     1
4721     1
4799     1
3089     1
Name: count, Length: 873, dtype: int64

In [67]:
CustomerAddress['customer_id'].nunique()

3999

In [114]:
df_left = pd.merge(CustomerDemographic, CustomerAddress, how='left',on='customer_id')
df_left.shape

(4000, 17)

In [115]:
df_inner = pd.merge(CustomerDemographic, CustomerAddress, how='inner',on='customer_id')
df_inner.shape

(3996, 17)

In [116]:
df_left.loc[df_left['address'].isna(),:]

customer_id first_name last_name  gender   
2             3      Arlin    Dearle    Male  \
9            10   Fiorenze   Birdall  Female   
21           22    Deeanne  Durtnell  Female   
22           23       Olav     Polak    Male   

    past_3_years_bike_related_purchases        DOB                job_title   
2                                    61 1954-01-20       Recruiting Manager  \
9                                    49 1988-10-11  Senior Quality Engineer   
21                                   79 1962-12-10                      NaN   
22                                   43 1995-02-10                      NaN   

   job_industry_category  wealth_segment deceased_indicator owns_car  tenure   
2               Property   Mass Customer                  N      Yes    15.0  \
9     Financial Services   Mass Customer                  N      Yes    20.0   
21                    IT   Mass Customer                  N       No    11.0   
22                   NaN  High Net Worth                  N      Yes     1.0   

   address  postcode state country  property_valuation  
2      NaN       NaN   NaN     NaN                 NaN  
9      NaN       NaN   NaN     NaN                 NaN  
21     NaN       NaN   NaN     NaN                 NaN  
22     NaN       NaN   NaN     NaN                 NaN

In [118]:
CustomerDemographic.query("customer_id == 4000")

customer_id first_name last_name gender   
3999         4000      Kippy   Oldland   Male  \

      past_3_years_bike_related_purchases        DOB             job_title   
3999                                   76 1991-11-05  Software Engineer IV  \

     job_industry_category     wealth_segment deceased_indicator owns_car   
3999                   NaN  Affluent Customer                  N       No  \

      tenure  
3999    11.0

In [121]:
df_right = pd.merge(CustomerDemographic, CustomerAddress, on='customer_id', how='right')
df_right.shape

(3999, 17)

In [122]:
df_right.loc[df_right['gender'].isna(),:]

customer_id first_name last_name gender   
3996         4001        NaN       NaN    NaN  \
3997         4002        NaN       NaN    NaN   
3998         4003        NaN       NaN    NaN   

      past_3_years_bike_related_purchases DOB job_title job_industry_category   
3996                                  NaN NaT       NaN                   NaN  \
3997                                  NaN NaT       NaN                   NaN   
3998                                  NaN NaT       NaN                   NaN   

     wealth_segment deceased_indicator owns_car  tenure   
3996            NaN                NaN      NaN     NaN  \
3997            NaN                NaN      NaN     NaN   
3998            NaN                NaN      NaN     NaN   

                     address  postcode state    country  property_valuation  
3996  87 Crescent Oaks Alley      2756   NSW  Australia                  10  
3997        8194 Lien Street      4032   QLD  Australia                   7  
3998         320 Acker Drive      2251   NSW  Australia                   7

=> Customers with id 3, 10, 22, 23 have CustomerDemographic informations but no CustomerAddress informations. Customers with id 4001, 4002, 4003 have CustomerAddress informations but no CustomerDemographic informations. 


In [68]:
CustomerAddress['state'].value_counts(dropna=False)

state
NSW                2054
VIC                 939
QLD                 838
New South Wales      86
Victoria             82
Name: count, dtype: int64

In [69]:
CustomerAddress['state'] = CustomerAddress['state'].replace(['New South Wales','Victoria'],['NSW','VIC'])

In [70]:
CustomerAddress['property_valuation'].value_counts(dropna=False)

property_valuation
9     647
8     646
10    577
7     493
11    281
6     238
5     225
4     214
12    195
3     186
1     154
2     143
Name: count, dtype: int64